async sensor API

yperess opened the pull request:
- Introduce a new data flow:
  - Request - data is requested and returned by the sensor asynchronously and without a syscall via RTIO
  - Raw - data is in a raw buffer format directly from the sensor driver
  - Processing - sensor drivers provide a decoder which will decode the data into a Q31 format which is compatible with zDSP (Zephyr Digital Signal Processing) subsystem.
- New API is fully backwards compatible. Benchmarking demonstrated being within **1%** performance of the exiting API when USERSPACE was disabled (it's likely that it might even beat it when userspace is enabled because no syscalls are required in the new API). Additionally, data is much more compressed in the raw data format while the data is waiting for processing.
- Using the RTIO enables the use of mempools which will allocate data as needed which allows several sensors to share the same context.
- Migrate the sensor shell to use the new API

yperess commented:
@teburd I got things working (kinda). I still can't get the thermistor to work side by side with the IMU (@asemjonovs do you have any thoughts?). But overall this is working really well. Let me know what you think.

teburd commented:
> @teburd I got things working (kinda). I still can't get the thermistor to work side by side with the IMU (@asemjonovs do you have any thoughts?). But overall this is working really well. Let me know what you think.

I'll take a closer look today. I guess at first glance I'm concerned about passing the rtio context to the sensor_read() call but I need to understand better why that might be needed by reading this.

yperess commented:
> > @teburd I got things working (kinda). I still can't get the thermistor to work side by side with the IMU (@asemjonovs do you have any thoughts?). But overall this is working really well. Let me know what you think.
> 
> I'll take a closer look today. I guess at first glance I'm concerned about passing the rtio context to the sensor_read() call but I need to understand better why that might be needed by reading this.

Yes, passing the context for every read was not my first choice, but it was the only way I could think of making this 100% backwards compatible. A workflow that does:
1. Set RTIO context
2. Trigger a read

Will not work without changes from the driver since we don't have anywhere to store the RTIO context. There are alternatives to create a meta data array of `{const struct device*, struct rtio *}` and we allocate one per sensor in an iterable section. I'm still open to this approach if you like it better.

yperess commented:
> Ok just to kind of re-dump what we had brainstormed...
> 
> ```c
> RTIO_DEFINE_WITH_MEMPOOL(r, 4, 4, BUF_POOL_COUNT, BUF_SIZE);
> 
> int main()
> {
>     const struct device *icm42688 = DEVICE_DT_GET(icm42688);
>     const struct device *akm09918 = DEVICE_DT_GET(akm09918);
> 
>     /* TODO Setup sensors including trigger source here (e.g. fifo en, watermark level, timers, whatever) */
>       
>     /* Setup a pool read for the sensors, using the sensor as the userdata */
>     sensor_read_with_pool(icm42688, r);
>     sensor_read_with_pool(akm09918, r);
>     
>     while (true) {
>       struct device *read_dev;
>       void *buf;
>       uint32_t len;
> 
>       /* Submit with wait count 1, immediately reads and deciphers the completion into the out params */
>       int res = sensor_read_await(r, &read_dev, &buf, &len);
> 
>       /* We now have a buffer with the sensor data or an error, its ours until we free it */
>       if (res < 0) {
>          LOG_ERR("failed to read sensor data %d\n", res);
>          continue;
>       } else {
>          if (read_dev == icm42888) {
>             process_icm42688(read_dev, buf, len);
>          }  else { 
>            process_akm09918(read_dev, buf, len);
>          }
>          sensor_read_free(r, buf, len);
>          sensor_read_with_pool(read_dev, r);
>       }
>       
>     }
> }
> ```

Correct, and what I have here is basically the same. The sample we have assumes FIFO and that the sensor is generating the data (which will come soon after this PR). What I have currently in this PR basically is a trigger one-shot read (`sensor_read()`) The while loop that you posted is basically what the extra thread in this PR is doing it's just that I didn't get around to doing the FIFO support so I needed 2 threads. One triggers and one processes data.
> 
> And honestly this doesn't seem too far off from that. I think the difference here is what maybe my mind was going to with the sensor_read() call was a wrapper around what amounts to the sensor acting like an I/O device that you can submit requests to, and sensor_read does all the setup for that for you.
> 
> So making sensor_read() a syscall I think is the the unexpected part in this setup. I was expecting there to be a driver level submit() type call (should not be a syscall) that takes the submission (has the rtio context as well!) to read and produces a completion.
> 
> I guess the difficulty is matching the idea that you'd like to read some N channels.
> 
> This can be encoded in the iodev's data and then the read call can take the iodev (configured with the channels and such). The read call doesn't need to be a syscall but can be an inline helper that creates the submission and submits it with the given context.

I think the difficulty here was actually making this backwards compatible (though maybe I just wasn't creative enough). My current thinking was that with the existing approach `sensor_read()` is a single syscall but it's replacing 2+ (fetch/get). For FIFO reading we would just add a new API after this `sensor_stream(const struct device *, struct rtio*)` if the RTIO context is NULL, then we'll turn off the stream, if it's non-NULL then we'll start streaming data to the same processing loop associated with the RTIO context.

> 
> ```c
> SENSOR_DT_IODEV(icm42688_gyros, sensor_dt_node, somechannels, num_channels)
> 
> sensor_read(icm42688_gyros, r);
> ```
> 
> Does that make sense or maybe I'm overlooking something else?
> 
> To make this backwards compatible then, a sensor_submit API is available that does kind of what you do here already...
> 
> ```c
> void sensor_iodev_submit(struct rtio_iodev_sqe *iodev_sqe) {
>   const struct sensor_iodev_data *iodev_data = iodev_sqe->sqe->iodev->data;
>   const struct device *sensor_dev = iodev_data->sensor_dev;
>   const struct sensor_driver_api *api = sensor_dev->api;
> 
>   /* Device does supports async reads with rtio */
>   if (api->submit != NULL) {
>     return api->submit(dev, iodev_sqe);
>   }
>   
>   /* submit not implemented, but presumably fetch and get are so use those to mimic submit */
>   enum sensor_channel *channels = iodev_data->channels;
>   const uint32_t channel_num = iodev_data->channel_num;
>   
>   sensor_sample_fetch(dev, SENSOR ...);
> 
>   /* for channel in channels */
>   for(int i = 0; i < channel_num; i++) {
>      sensor_channel_get(dev, ....);
>   }
> 
>   /* Return completion... */
>   rtio_cqe_submit(iodev_sqe->r, cqe, ...);
> }
> ```

I was aiming for abstracting away the RTIO details though I'll admit it comes at the cost of granularity. If the developer wants to do a single read with a pre-allocated buffer to a mempool RTIO, they will not be able to with my API, though we could add a buffer/size argument and allow the developer to use `NULL` if the context has a mempool. That wouldn't be too difficult.

> 
> I do kind of like that you've made the processing take a function pointer so cleanup and everything is automatic. This too should probably be a helper function kind of thing. So rather than sensor_read_await you you call sensor_read_process(r, process_func) kind of thing perhaps?
> 
> Anyways, I think this is really on the right track, but maybe some missed thoughts on my part weren't conveyed well. I hope that helps!

teburd commented on include/zephyr/rtio/rtio.h:
```diff
@@ -1075,9 +1051,8 @@ static inline int rtio_sqe_rx_buf(const struct rtio_iodev_sqe *iodev_sqe, uint32
-				map_entry->state = RTIO_MEMPOOL_ENTRY_STATE_ALLOCATED;
-				map_entry->block_idx = __rtio_compute_mempool_block_index(r, *buf);
-				map_entry->block_count = num_blks;
+				mutable_sqe->buf = *buf;
```
I'm wondering what happened with the mempool situation here, seems like a complete change to what you had?

yperess replied:
Something isn't quite right, it is not working well with high data rates.

teburd reviewed (commented):
This is looking really good to me, some commentary on the mempool changes would be good

yperess commented:
> This is looking really good to me, some commentary on the mempool changes would be good

Something isn't quite right, it is not working well with high data rates. I'll try to debug tonight.

yperess commented:
@teburd, performance is abysmal currently (see output below). The fact that we're only getting 512 callbacks though makes me wonder if this is an issue with the CQE not being released properly?

```
Statistics:
Duration: 0.500
Num Requests: 36914
Sampling frequency: 73828Hz (ideal: 32000Hz)
Reading frequency: 512Hz
Num Callbacks: 512
Num Samples: 256
Drop rate: 50.00
Num Failed Reads: 256
Num Failed Timestamps: 0
Num Failed Frame Counts: 0
Num Failed Get Scale: 0


*************************
Legacy Statistics:
Duration: 0.500
Sampling frequency: 80582Hz (ideal: 32000Hz)
Reading frequency: 31658Hz
Num Requests: 40291
Num Samples: 15829
```

teburd commented:
I don't see what you are seeing when I run this, so I'm not sure what we're measuring here

When I run this branch as is, zero samples are missed and the polling sample loop works at 3KHz  just fine as seen on the logic analyzer.



![Screenshot from 2023-04-25 08-15-11](https://user-images.githubusercontent.com/135839/234288212-db31c04c-94cd-448e-8d50-50b4f8675258.png)

teburd commented:
Ok correction, I do see what you are seeing now, but lets look at whats on the logic analyzer.

Between spi reads chip select times is 30us, in the polling loop. This is 50KHz polling over a spi bus that itself uses a tight polling loop. Meaning in effect zero computation time is available, that other 20us is the cost of doing business.

![Screenshot from 2023-04-25 08-21-02](https://user-images.githubusercontent.com/135839/234290338-4daef887-67eb-4e05-94d4-e4b9285a9030.png)

Then at a larger scale there's a large bank of time where no spi work is done, nearly a second in fact. So something isn't quite right there. Your processing thread isn't being scheduled in in between reads like I think you want it to be is that right?

![Screenshot from 2023-04-25 08-23-01](https://user-images.githubusercontent.com/135839/234290272-7f217efb-453f-47f1-97b9-45d88be917dc.png)

That seems like a different issue. The throughput of queueing reads and getting completions seems to be quite good to me.

yperess commented:
@teburd @MaureenHelm I believe this is ready for first pass review (the sample used here is a hacked sensor_shell and that final commit will be replaced tomorrow by actually updating the sensor shell implementation to use this API instead. Some metrics from the TDK Robokit1 board:

When running the sensor at 32kHz, and using a tight sampling loop:

**Existing fetch/get APIs**
31,678Hz (1.00625% of the samples were dropped) and used 48 bytes for 6 channels of data with 1&#x00B5;m/s&#x00B2; precision for the accelerometer and 1&#x00B5;rad/s for the gyroscope

**Async API**
31330Hz (2.09375% of the samples were dropped) and used 48 bytes for 6 channels of data + 64bit timestamp with a worst case precision of 0.073&#x00B5;m/s&#x00B2; and 0.0016&#x00B5;rad/s for the accelerometer and gyroscope respectively.

**What's next?**
- Update the sample by actually implementing the sensor_shell `get` API with the new API.
- Post a second PR to demonstrate the next layer which is the streaming support. It will build ontop of these changes.

yperess commented on include/zephyr/drivers/sensor.h:
```diff
@@ -394,12 +398,171 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
+	 * @return 0 on success
+	 * @return <0 on error
+	 */
+	int (*get_timestamp)(const uint8_t *buffer, uint64_t *timestamp_ns);
```
@teburd @MaureenHelm I didn't think that ms resolution was good enough but I understand that many boards will not have ns resolution either.
Design choice:
- [ ] Use microseconds
- [X] Use nanoseconds
- [ ] Use ticks

I went with nanoseconds, but honestly I'm leaning towards just ticks and letting the user compute it since there are plenty of helpers in kernel.h

yperess replied:
Updated to ticks

yperess commented:
@teburd @MaureenHelm Updates:

1. timestamp is now ticks
2. `uint32_t scale` is now an `int8_t shift` so data compression is better (reading accel_xyz + gyro_xyz uses a smaller buffer than the existing APIs). Also, using the shift instead of a free form scale makes for faster math
3. Updated the sensor_shell.c to use the new API which worked out a few bugs when dynamically configuring the read iodev.

Sample:
```
uart:~$ sensor get icm42688p@0 accel_xyz gyro_xyz die_temp
channel idx=0 accel_x value=1.626859
channel idx=1 accel_y value=1.321598
channel idx=2 accel_z value=5.795164
channel idx=4 gyro_x value=0.002663
channel idx=5 gyro_y value=-0.018384
channel idx=6 gyro_z value=0.005594
channel idx=12 die_temp value=28.313706
```

teburd commented on drivers/sensor/default_rtio_sensor.c:
```diff
@@ -0,0 +1,371 @@
+ * @brief Default decoder get the timestamp of the first frame
+ *
+ * @param[in]  buffer The data buffer to parse
+ * @param[out] timestamp_ticks The timestamp of the first frame
```
I'd be wary of using ticks here, I actually think the timestamp_ns is smarter as the clock source could be more precise than ticks.

Ticks to ns is possible, so nothing else would need to change but the signature of this for me to +1 today

yperess replied:
Done

teburd reviewed (changes requested):
I think the timestamp you had right before in using a precision time unit (nanoseconds) as opposed to ticks.

There's almost always going to be a higher rate clock available to get more precise changes in time than a tick.

Everything else I think is looking really good. I'd remove the z_ from the sensor_processing_loop helper as well :-)

+1 from me given those changes

yperess commented on include/zephyr/drivers/sensor.h:
```diff
@@ -394,12 +398,184 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
+	};                                                                                         \
+	RTIO_IODEV_DEFINE(name, &__sensor_iodev_api, &__sensor_read_config_##name)
+
+static inline void sensor_reconfigure_read_iodev(struct rtio_iodev *iodev,
```
@teburd I want to get your take on this function, it is not currently used by the sensor_shell but it i think it needs to be when userspace is enabled.

teburd replied:
This seems fine, but would need to be made a syscall when you enable userspace.

yperess replied:
Correct

yperess replied:
Done

yperess replied:
@teburd I enabled userspace and ran the sensor shell sample with directly manipulating the array and it worked. My guess is that shell commands run in supervisor mode. Any thoughts of a use case we can commit for this?

yperess commented on drivers/sensor/sensor_shell.c:
```diff
@@ -113,6 +116,20 @@ enum dynamic_command_context {
+RTIO_IODEV_DEFINE(iodev_sensor_shell_read, &__sensor_iodev_api, &iodev_sensor_shell_read_config);
+
+/* Create the RTIO context to service the reading */
+RTIO_EXECUTOR_SIMPLE_DEFINE(sensor_shell_executor);
```
Right, I want to add a new sensor shell command in the next PR `sensor stream icm42688 <trigger_0>..<trigger_N>` and since that one will be "always on" I thought we would want to use the concurrent executor. For the existing example the shell seems to be more of a bottleneck than the API so I went with what's simple to avoid overloading too many changes into the same PR.

yperess commented on include/zephyr/math/ilog2.h:
```diff
@@ -93,7 +93,7 @@
 	(						\
 		__builtin_constant_p(n) ?		\
-		ilog2_compile_time_const_u32(n) :	\
+		ilog2_compile_time_const_u32((n)) :	\
```
Done

teburd commented on include/zephyr/drivers/sensor.h:
```diff
@@ -394,12 +398,184 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
+ *
+ * @see sensor_get_decoder for more details
+ */
+typedef int (*sensor_get_decoder_t)(const struct device *dev,
```
I think this is ok if some rules are applied. The API struct needs to be global (not on a stack somewhere), and needs to be viewed as immutable from the user space/callers perspective.

If those hold true this should be fine. Alternatively you could copy it to caller provided memory. The benefits of the copy might be there isn't an unexpected mutation from the driver causing some hassles here.

``` c
struct sensor_decoder_api decoder;

sensor_get_decoder(sensor_dev, &decoder); 
```

doesn't seem too bad.

yperess replied:
Right, the goal (not yet documented) is to have these just 1 per sensor and they should always be the exact same from boot (drivers should never mutate them _EVER_).

teburd replied:
In any case after discussing with @ceolin this does need to copy out

yperess replied:
Done

yperess commented:
Next steps:
- [ ] Add rst documentation
- [ ] In separate PRs, add implementations for ICM42688 and AKM09918 along with updating the tests (@asemjonovs @tristan-google respectively)
- [ ] In separate PR, add streaming APIs that use the exact same dataflow.

This PR should remain with the DNM label until the above are provided.

yperess commented on subsys/rtio/rtio_executor_simple.c:
```diff
@@ -71,13 +73,26 @@ int rtio_simple_submit(struct rtio *r)
+		rtio_spsc_release(r->sq);
+	} else if (FIELD_GET(RTIO_SQE_MEMPOOL_BUFFER, sqe->flags) == 1 && sqe->op == RTIO_OP_RX) {
+		/* Multi-shot read with mempool buffer, reset the memory for the next read */
+		LOG_DBG("Multi-shot with mempool detected, clearing buffers for %p", (void *)sqe);
+		sqe->buf = NULL;
+		sqe->buf_len = 0;
```
@teburd I'm having some difficulties with this one. So we have the flag for multishot and I want to allow other tasks to run which means really what I want to do is put this at the end of the `sq`, but... I can't really do that because then I'll have a new sqe handle for the `rtio_sqe_cancel()` operation. So I'm wondering if the solution is to make the `rtio_spsc_consume()` a bit more complicated by having the fallback (when nothing was found at HEAD) to scan the entire thing for valid SQEs. I don't like this solution, but I'm also not really seeing how we can avoid fragmentation with multishot support.

teburd replied:
I have a branch that resolves the issue with spsc head of line blocking here. It drops the rings (spsc queues) in favor of lists everywhere, so for getting a sqe, you pull from a free list, for pushing to the submission queue, its a mpsc_push, and so on.

I'll push it up later today, we can work out the ordering of the commits and such then.

yperess commented on include/zephyr/rtio/rtio.h:
```diff
@@ -1256,12 +1281,36 @@ static inline int z_impl_rtio_sqe_copy_in(struct rtio *r,
+ */
+static inline int rtio_sqe_copy_in(struct rtio *r, const struct rtio_sqe *sqes, size_t sqe_count)
+{
+	return z_impl_rtio_sqe_copy_in_get_handles(r, sqes, NULL, sqe_count);
```
Fixed, waiting on @asemjonovs to wrap up the one shot implementation for the icm42688 so I can add the streaming support to the same API submit function.

yperess commented:
@teburd It's really really close but I can't seem to get the CQE in the sensor shell. If you build and run the sensor shell sample, then run `sensor stream cim42688p@0 on fifo_full` you'll see that we read the FIFO but 2 things happen (or don't):
1. It's not generating the CQE that would unblock the processing thread.
2. It doesn't seem to re-issue the SQE to the iodev (I believe that's because of the CQE issue).

nashif commented on west.yml:
```diff
@@ -22,16 +22,9 @@ manifest:
     - name: upstream
       url-base: https://github.com/zephyrproject-rtos
 
-  group-filter: [-babblesim]
```
wonder why this is needed :)

yperess replied:
I haven't been able to run west or twister without this revert

Commit:
sensors: Add streaming APIs

Introduce a streaming API that uses the same data path as the async API.

This includes features to the decoder:
* Checking if triggers are present

Adding streaming features built ontop of existing triggers:
* Adding 3 operations to be done on a trigger
  * include - include the data with the trigger information
  * nop - do nothing
  * drop - drop the data (flush)
* Add a new sensor_stream() API to mirror sensor_read() but add an
optional handler to be able to cancel the stream.

Signed-off-by: Yuval Peress <peress@google.com>
topic#sensor_stream

Commit:
Rework sensing subsystem

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
Add open/close support

Commit:
Implement arbitration

Commit:
Add sensing shell

Commit:
add support for userspace

Commit:
Implement pipe

Commit:
icm42688: Implement streaming APIs

Add streaming implementation for icm42688 using both threshold and
full FIFO triggers.

Signed-off-by: Yuval Peress <peress@google.com>
topic#sensor_stream

Commit:
Implement angle sensor

Commit:
Rework logic to encode the data in the drivers
